### PR TITLE
Off-chain Messages

### PR DESCRIPTION
**Pull request by JulianToledano**

### CODE REVIEW

File: cips/cip-X.md [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+      }
+    ],
+    "memo": ""
+  }
+}
+```
+
+## Drawbacks
+
+This CIP requires a fixed relationship between an account address and a public key. That means it won't work if [ADR-034](https://github.com/cosmos/cosmos-sdk/blob/v0.50.0-rc.1/docs/architecture/adr-034-account-rekeying.md)
```
Comment (tac0turtle) [inline cips/cip-X.md:218]: why is this required?

File: cips/cip-X.md [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+
+There was a prior attempt to implement this in the sdk with [ADR-036](https://github.com/cosmos/cosmos-sdk/blob/v0.50.0-rc.1/docs/architecture/adr-036-arbitrary-signature.md).
+However, a consensus was not achieved on ADR-036, as the community demanded a way for sign-in method that ADR-036
+did not provide. This is the main reason for defining two different messages in this CIP. One for arbitrary data and
+the other for sing-in.
+
+
+## Unresolved Questions
+
+ + Is defining different messages per use case the best approach, or could a protobuf with a `oneof` be sufficient?
```
Comment (tac0turtle) [inline cips/cip-X.md:256]: i dont think either need to happen. The message can be encoded and added to the msgarbitray signing right?

File: cips/cip-X.md [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+## Specification
+
+Off-chain signed messages should resemble Cosmos SDK messages but **must not** constitute a valid on-chain transaction.
+`chain-id`, `account_number`, and `sequence` must all be assigned invalid values.
+
+An off-chain transaction follows these rules:
+
+ + the memo must be empty
+ + nonce, sequence number must be equal to 0
+ + chain-id must be equal to “”
```
Comment (hxrts) [inline cips/cip-X.md:77]: non-empty chain-id might be helpful if you want to bind the signed object to a specific chain

File: cips/cip-X.md [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+offers a more user-friendly sign-in process, replacing non-human-readable random challenges.
+
+## Specification
+
+Off-chain signed messages should resemble Cosmos SDK messages but **must not** constitute a valid on-chain transaction.
+`chain-id`, `account_number`, and `sequence` must all be assigned invalid values.
+
+An off-chain transaction follows these rules:
+
+ + the memo must be empty
```
Comment (hxrts) [inline cips/cip-X.md:74]: I could see a bunch of use-cases for non-empty memo

File: cips/cip-X.md [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+ + nonce, sequence number must be equal to 0
+ + chain-id must be equal to “”
+ + fee gas must be equal to 0
+ + fee amount must be an empty array
```
Comment (hxrts) [inline cips/cip-X.md:79]: gas fee, fee amount you may want to set to -1 to properly distinguish them from on-chain txs since it’s possible to send zero fee txs

File: cips/cip-X.md [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+  }
+}
+```
+
+## Drawbacks
+
+This CIP requires a fixed relationship between an account address and a public key. That means it won't work if [ADR-034](https://github.com/cosmos/cosmos-sdk/blob/v0.50.0-rc.1/docs/architecture/adr-034-account-rekeying.md) 
+is implemented.
+
+Doesn't work with multisig accounts.
```
Comment (hxrts) [inline cips/cip-X.md]: is this true? aren't all multi-sig pubkeys posted in a deterministic fashion now so the mapping is unique

File: cips/cip-X.md (hunk 2 above)
Comment (JulianToledano) [reply cips/cip-X.md:256]: Problem with that is ledger, as it won't be readable for users

File: cips/cip-X.md (hunk 5 above)
Comment (JulianToledano) [reply cips/cip-X.md:79]: I have to check if thats possible

File: cips/cip-X.md (hunk 1 above)
Comment (JulianToledano) [reply cips/cip-X.md:218]: Here is some context about this:
https://github.com/cosmos/cosmos-sdk/pull/7727#issuecomment-723966597

Comment (tac0turtle) [reply cips/cip-X.md:218]: can you explain in the rfc why this is needed? @testinginprod do know of any ways to avoid this?

File: cips/cip-X.md [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+author: @JulianToledano
+status: Draft
+type: Standards Track
+created: 2023-10-10
+---
+
+## Simple Summary
+
+This is a convention for signing arbitrary messages. We propose this specification as a method for signing and
+validating off-chain arbitrary messages.
```
Comment (webmaster128) [inline cips/cip-X.md]:
```suggestion
verifying off-chain arbitrary messages.
```

File: cips/cip-X.md [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+format that is parameterized by scope, session details, and security mechanisms.
+
+## Motivation
+
+Having the ability to sign messages off-chain has proven to be a fundamental aspect of nearly any blockchain.
+The notion of signing messages off-chain has many added benefits, such as saving on computational costs and reducing
+transaction throughput and overhead. Within the context of the Cosmos ecosystem, some of the major applications of signing such
+data include, but are not limited to, providing a cryptographically secure and verifiable means of proving validator
+identity and possibly associating it with some other framework or organization. Additionally, having the ability to
+sign Cosmos messages with a Ledger or similar HSM device.
```
Comment (webmaster128) [inline cips/cip-X.md]: You could add smart accounts here, where off-chain signing allows for application layer authorizations of something.

E.g. we have a CosmWasm smart contract that verifies Ethereum arbitrary messages in a contract for a long time. Similarly folks now request secp256r1 message verification in CosmWasm for the same use case: some keypair signs arbitrary data off-chain and the contract can verify it.

File: cips/cip-X.md [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+
+This approach is an improvement over the current state, as it allows for the signing of arbitrary data like files and
+offers a more user-friendly sign-in process, replacing non-human-readable random challenges.
+
+## Specification
+
+Off-chain signed messages should resemble Cosmos SDK messages but **must not** constitute a valid on-chain transaction.
+`chain-id`, `account_number`, and `sequence` must all be assigned invalid values.
+
+An off-chain transaction follows these rules:
```
Comment (webmaster128) [inline cips/cip-X.md:73]: What's the rule for the account number? Account number 0 is valid.

File: cips/cip-X.md [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+  "nonce": "14368412",
+  "issued-at": "2023-10-10T10:10:10Z"
+```
+
+This approach is an improvement over the current state, as it allows for the signing of arbitrary data like files and
+offers a more user-friendly sign-in process, replacing non-human-readable random challenges.
+
+## Specification
+
+Off-chain signed messages should resemble Cosmos SDK messages but **must not** constitute a valid on-chain transaction.
```
Comment (webmaster128) [inline cips/cip-X.md]:
```suggestion
Off-chain signed messages should resemble Cosmos SDK transactions but **must not** constitute a valid on-chain transaction.
```

File: cips/cip-X.md (hunk 3 above)
Comment (webmaster128) [reply cips/cip-X.md:77]: I see the use case but don't think this belongs here. No chain ID is good for _not valid on any chain_. Any app specific metadata should go into the messages.

File: cips/cip-X.md (hunk 5 above)
Comment (webmaster128) [reply cips/cip-X.md:79]: If the transaction becomes invalid through other restrictions above, we can use a valid fee value here

File: cips/cip-X.md [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+ + the memo must be empty
+ + nonce, sequence number must be equal to 0
+ + chain-id must be equal to “”
+ + fee gas must be equal to 0
+ + fee amount must be an empty array
+
+Verification of an off-chain transaction follows the same rules as an on-chain one, except for the specification
+differences highlighted above.
+
+As messages in cosmos are defined as proto definitions, different messages can be defined for different off-chain
```
Comment (webmaster128) [inline cips/cip-X.md]:
```suggestion
As messages in Cosmos are defined as proto definitions, different messages can be defined for different off-chain
```

File: cips/cip-X.md [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+
+`MsgSignIn` enables the proof of wallet ownership for applications sign-in. In this context, `AppDomain` is the
+application requesting off-chain message signing. `URI` refers to the resource that is the subject of the signing.
+`signerAddress` is the account address of the signer. `Nonce` is a random string typically chosen by the relying on party and
+used to prevent replay attacks. `Issued-at` the time when the message was generated.
+
+Proto definition:
+
+```protobuf
+// MsgSignArbitraryData defines an arbitrary, general-purpose, off-chain message
```
Comment (webmaster128) [inline cips/cip-X.md]:
```suggestion
// MsgSignIn defines an arbitrary, general-purpose, off-chain message
```

File: cips/cip-X.md [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+There was a prior attempt to implement this in the sdk with [ADR-036](https://github.com/cosmos/cosmos-sdk/blob/v0.50.0-rc.1/docs/architecture/adr-036-arbitrary-signature.md).
+However, a consensus was not achieved on ADR-036, as the community demanded a way for sign-in method that ADR-036
+did not provide. This is the main reason for defining two different messages in this CIP. One for arbitrary data and
+the other for sing-in.
+
+
+## Unresolved Questions
+
+ + Is defining different messages per use case the best approach, or could a protobuf with a `oneof` be sufficient? 
+ + Should `Data` in `MsgSignArbitraryData` have a max length?
```
Comment (webmaster128) [inline cips/cip-X.md:257]: No, but it should be clear that not every signer is capable of signing arbitrary length messages

File: cips/cip-X.md [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+
+Verification of an off-chain transaction follows the same rules as an on-chain one, except for the specification
+differences highlighted above.
+
+As messages in cosmos are defined as proto definitions, different messages can be defined for different off-chain
+use cases, such as Sign-In, proof of wallet ownership or the ability to sign arbitrary data.
+
+All off-chain messages will be of the type `offchain/messageKind`.
+
+### MsgSignArbitraryData
```
Comment (webmaster128) [inline cips/cip-X.md:89]: I think it would be a missed opportunity to not start with arbitrary text here. We'll present data to the user that cannot be printed or verified, making it blind signing by default.

File: cips/cip-X.md [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+        },
+        "signature": "8y8i34qJakkjse9pOD2De+dnlc4KvFgh0wQpes4eydN66D9kv7cmCEouRrkka9tlW9cAkIL52ErB+6ye7X5aEg=="
+      }
+    ],
+    "memo": ""
+  }
+}
+```
+
+### MsgSignIn
```
Comment (webmaster128) [inline cips/cip-X.md]: Shouldn't this be something like `MsgProveIdentity`? I wonder if this spec knows how all the login stuff works.

File: cips/cip-X.md [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+  // Signer is the sdk.AccAddress of the message signer
+  string signerAddress = 3 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+  // Nonce is a random string
+  string nonce = 4;
+  // Issued-at is the time when the message was generated
+  google.protobuf.Timestamp issued-at = 5;
+}
+```
+
+Signed MsgSignIn json example:
```
Comment (webmaster128) [inline cips/cip-X.md]:
```suggestion
Signed MsgSignIn JSON example:
```

File: cips/cip-X.md [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+
+
+## Unresolved Questions
+
+ + Is defining different messages per use case the best approach, or could a protobuf with a `oneof` be sufficient? 
+ + Should `Data` in `MsgSignArbitraryData` have a max length?
+
+## Backwards Compatibility
+
+Backwards compatibility is guaranteed as this CIP introduces new messages but does not modify any prior work.
```
Comment (webmaster128) [inline cips/cip-X.md]: I think it's worth mentioning that it is incompatible to ADR-036 as it defines different message types

**Review by webmaster128 — COMMENTED** [review]
🍾❤️🚀

File: cips/cip-X.md (hunk 1 above)
Comment (webmaster128) [reply cips/cip-X.md:218]: I think for stateless verification it makes sense to keep this restriction (and not implement ADR-034)

Comment (tac0turtle) [reply cips/cip-X.md:218]: makes sense for stateless. adding a few explanation lines in the paragraph would be useful

File: cips/cip-X.md (hunk 2 above)
Comment (tac0turtle) [reply cips/cip-X.md:256]: for this usecase i think thats fine. We dont want to overcomplicate the system. The user will know whats being signed in the browser and would be able to verify the humanreadable bytes match the client

**Comment by JulianToledano** [comment]
@tac0turtle @webmaster128 
Hey guys! I did commit most of your suggestions. One thing I don't have clear is:
should we go with just one message or both?
As commented above the main problem with ADR-036 was that the message defined wasn't human readable.

File: cips/cip-X.md (hunk 6 above)
Comment (tac0turtle) [reply cips/cip-X.md]: it should work with multisigs as the current multisig is deterministic, but on chain multisigs would not work here. It might make sense to clarify this.

**Comment by webmaster128** [comment]
> should we go with just one message or both?

I like the separate MsgProveIdentity. Sounds specific and flexible at the same time.

For the arbitrary content signing I'd go with arbitrary valid UTF-8 text. This is readable by default, encourages the design of readable solutions and can host arbitrary data as well. E.g. any JSON document is a printable string and embedding raw data using some kind of encoding is just as good as a `bytes` blob that the UI has to show as base64 anyways. Given that I don't see a reason for arbitrary non-string data signing.

File: cips/cip-X.md (hunk 6 above)
Comment (tac0turtle) [reply cips/cip-X.md]: gentle ping on adding some verbiage here

File: cips/cip-X.md [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+
+This CIP defines two different messages. The first one allows users to sign any arbitrary data, including files, text,
+objects, and more. The second one provides a way for application sign-in by defining a standard human-readable message
+format that is parameterized by scope, session details, and security mechanisms.
+
+## Motivation
+
+Having the ability to sign messages off-chain has proven to be a fundamental aspect of nearly any blockchain.
+The notion of signing messages off-chain has many added benefits, such as smart accounts where off-chain signing allows 
+for application layer authorizations saving on computational costs and reducing  transaction throughput and overhead.
```
Comment (tac0turtle) [inline cips/cip-X.md]:
```suggestion
for application layer authorizations saving on computational costs and reducing transaction throughput and overhead.
```

**Review by tac0turtle — APPROVED** [review]

File: cips/cip-X.md [hunk 19]
```diff
@@ -262,9 +262,6 @@ Backwards compatibility is guaranteed as this CIP introduces new messages but do
 
 ## Security Considerations
 
-It is important to notice that the `Data` field in `MsgSignArbitraryData` consists of bytes, which means it could be
-non-human-readable.
```
Comment (webmaster128) [inline cips/cip-X.md]: I think this can be rephrased but can remain here. Valid UTF-8 string does not mean
1. the string is printable
2. the signing device is capable of printing it

Examples of such valid strings are NULL, [BELL](https://en.wikipedia.org/wiki/Bell_character), emojis or unicode in the [Private Use Areas](https://en.wikipedia.org/wiki/Private_Use_Areas).

**Review by testinginprod — APPROVED** [review]
LGTM.

I would sync with Keplr since I saw they have a MsgSignData they use for off-chain signing, and understand how this might impact them.

Thank you.

**Comment by tac0turtle** [comment]
merging for now, but we will still sync with wallets and clients

**Comment by dogemos** [comment]
one minor feedback from @thunnini.

could an optional chainID field be added?

**Comment by webmaster128** [comment]
> could an optional chainID field be added?

This was discussed somwhere in here. Where and why do you need this? Since off-chain messages are outside of any chain, I don't see how this belongs in here. In MsgSignArbitraryData you can add application specific context (which might include a chain ID). MsgProveIdentity proved a keypair ownership independent of a chain.

**Comment by JulianToledano** [comment]
> > could an optional chainID field be added?
> 
> This was discussed somwhere in here.

[Here](https://github.com/cosmos/cips/pull/22#discussion_r1360474375) is the discussion.
